sunscreen_math/src/ring/barret.rs

```diff
@@ -0,0 +1,299 @@
+use std::{marker::PhantomData, ops::Not};
+
+use crypto_bigint::{
+    subtle::{ConditionallySelectable, ConstantTimeLess},
+    Uint,
+};
+
+use super::ArithmeticBackend;
+
+/// Contains precomputed values needed for a Barrett reduction in
+/// a ring Z_q
+///
+/// # Remarks
+/// This algorithm is only guaranteed to work so long as Modulus fits into a 64 * N - 1 bit
+/// value.
+pub trait BarretConfig<const N: usize> {
+    /// The modulus defining the ring
+    const MODULUS: Uint<N>;
+
+    /// floor(2**(64*N) / MODULUS)
+    const R: Uint<N>;
+
+    /// floor(2**(128*N) / MODULUS) - 2**(64*N) * R
+    const S: Uint<N>;
+
+    /// 2**(64*N) - R * MODULUS
+    const T: Uint<N>;
+}
+
+pub struct BarretBackend<const N: usize, C: BarretConfig<N>> {
+    _phantom: PhantomData<C>,
+}
+
+impl<const N: usize, C: BarretConfig<N>> BarretBackend<N, C> {
+    /// Compute x (a 2N limb value) mod C::MODULUS
+    ///
+    /// # Remarks
+    /// See https://math.stackexchange.com/questions/3455277/barrett-reduction-possible-without-overflow-and-floating-point-arithmetic
+    ///
+    /// A key observation is that x = x_lo + 2^(64*N)*x_hi and thus
+    /// x mod m = x_lo mod m + 2^(64*N)*x_hi.
+    /// This is how we derive the required values in [`BarretConfig`], but the full
+    /// derrivation is in the link.
+    ///
+    /// We have carefully chosen the terms to obviate shifting and we can simply do
+    /// mulhi with no shifting.
+    fn barret_reduce(x: (Uint<N>, Uint<N>)) -> Uint<N> {
+        let (x_lo, x_hi) = x;
+
+        fn reduce<const N: usize, C: BarretConfig<N>>(val: &mut Uint<N>) {
+            let reduced = val.wrapping_sub(&C::MODULUS);
+
+            val.conditional_assign(&reduced, val.ct_lt(&C::MODULUS).not());
+        }
+
+        // Compute `mod_hi = x_hi * T - x_hi * s * p`
+        let asp = x_hi.mul_wide(&C::S).1.wrapping_mul(&C::MODULUS);
```

Comment:
This is *the* devil magic trick that makes Barrett reductions work. The idea is you want to divide by n without doing division. So you create a constant `r=2**k/n` and you can compute `rx >> k`. So long as k is large enough, `rx == x / n` with of an error of at most 1.

Now for the second trick where you bring Baal and Beelzebub together. If k == word size, then you can simply do mulhi(r, x), since this upper result is already shifted down. Hence, we ignore the lower word!